doc/developer/platform/architecture-storage.md

```diff
@@ -107,25 +107,150 @@ All commands are durably persisted and their effect survives crashes and reboots
+
+The goal of the source ingestion pipeline is to produce data that can be appended to the `persist` shards of a durable storage collection.
+The result of this collection is the primary user-facing output of the source ingestion pipeline.
+There may be other auxiliary information, like the reclocking collection, that feeds in to health checks and user-facing diagnostics.
+
+We want source ingestion to satisfy at least these three properties:
+
+1. Definiteness: this property is provided by `persist`.
+2. "Fault tolerance": the source should produce specific output even in the presence of unspecified i. failures and restarts, ii. concurrency with other source implementations.
+3. "Bounded input reliance": the raw input data we have to keep around to provide these properties should be as limited as possible.
+
+## A recipe for source ingestion
+
+What follows is a recipe for source ingestion that means to provide the three properties listed above.
+
+In addition, the "recipe" aspect of this means to make it "simple" to determine that you have these properties.
+You could certainly hand roll something that also has the properties, but then you have a bunch of work to do to convince folks.
+By following the recipe, the goal is that you shouldn't have to convince anyone of anything.
+
+### Ingredients
+
+1. We have some "raw sources" of data.
+
+    These sources implement something like our `SourceReader` trait, which has a method `get_next_message()` that produces an ordered sequence of messages.
+    Messages come with some ordered "address" (e.g. a (part, offset) pair or a LSN position or byte offset).
+    The messages are emitted in an order that matches the address (the address never goes backwards).
+    The source can be initialized at an address.
+    For each address, the source produces the same message to all readers (or "NA" if it's gone).
+
+2. We have access to the `persist` framework that can record evolving collections
+
+    The framework accepts arbitrary `Append` statements to individual shards and confirm when and whether they are durably recorded.
+    It handles contended writes, picking at most one winner, and providing a single consistent history of updates for each shard.
+    A collection may be formed of multiple shards, which may not have the same durable frontiers.
+
+3. We have an evolving understanding of what should happen to our data from raw sources to `persist`.
+
+    We want some amount of flexibility here, as we may be surprised by the requirements of source decoding, or new opportunities.
+    The "understanding" should reflect deterministic computation.
+    Non-determinism is something we can do with an advanced recipe (more `persist` stuff, in the spirit of reclocking, where we record non-deterministic choices).
+
+### The recipe
+
+There is some prep work to do first:
+
+1.  Identify all "raw sources" that contribute to some intended output (e.g. kafka topics, metadata, schema registry).
```

Comment:
I think I used "_base_ source" in my "descriptive" docs, but "raw source" seems like a better name!